File: Sources/Pomolego/Models/World.swift
import Foundation

struct GridCell: Hashable, Codable, Equatable {
    var col: Int
    var row: Int
}

struct PlacedBlock: Codable, Equatable, Identifiable {
    var id: UUID = UUID()
    var col: Int
    var row: Int
    var designID: String
    var isCracked: Bool
    var placedAt: Date

    var cell: GridCell { GridCell(col: col, row: row) }
}

/// The persistent 2D block world. Row 0 is the ground (bottom) row;
/// rows grow upward.
struct World: Codable, Equatable {
    static let columns = 28
    static let rows = 14

    var blocks: [PlacedBlock] = []

    private var occupancy: Set<GridCell> {
        Set(blocks.map(\.cell))
    }

    func block(at cell: GridCell) -> PlacedBlock? {
        blocks.first { $0.col == cell.col && $0.row == cell.row }
    }

    func isOccupied(_ cell: GridCell) -> Bool {
        block(at: cell) != nil
    }

    func inBounds(_ cell: GridCell) -> Bool {
        (0..<World.columns).contains(cell.col) && (0..<World.rows).contains(cell.row)
    }

    /// Free placement: any in-bounds, unoccupied cell is valid. Blocks may
    /// float anywhere on the grid — there is no gravity/support requirement.
    func isValidPlacement(_ cell: GridCell) -> Bool {
        inBounds(cell) && !isOccupied(cell)
    }

    /// Every empty cell on the grid.
    var validCells: [GridCell] {
        let occupied = occupancy
        var result: [GridCell] = []
        for col in 0..<World.columns {
            for row in 0..<World.rows {
                let cell = GridCell(col: col, row: row)
                if !occupied.contains(cell) { result.append(cell) }
            }
        }
        return result
    }

    /// Highest occupied row in a column (+1), used only for the idle menu-bar
    /// skyline glyph. With free placement this is just "tallest filled row".
    func columnHeight(_ col: Int) -> Int {
        let rows = blocks.filter { $0.col == col }.map { $0.row }
        guard let top = rows.max() else { return 0 }
        return top + 1
    }

    var mostRecentBlock: PlacedBlock? {
        blocks.max(by: { $0.placedAt < $1.placedAt })
    }

    /// Default target when the user starts without picking a spot: directly
    /// above the most recently placed block (so consecutive sessions stack
    /// naturally) if that cell is free, otherwise a free cell near the
    /// bottom-center of the grid.
    func defaultTarget() -> GridCell? {
        if let recent = mostRecentBlock {
            let above = GridCell(col: recent.col, row: recent.row + 1)
            if isValidPlacement(above) { return above }
        }
        return nearestValidGroundCell(to: World.columns / 2) ?? validCells.first
    }

    func nearestValidGroundCell(to col: Int) -> GridCell? {
        validCells
            .filter { $0.row == 0 }
            .min(by: { abs($0.col - col) < abs($1.col - col) })
    }

    @discardableResult
    mutating func place(designID: String, isCracked: Bool, at cell: GridCell, date: Date) -> Bool {
        guard isValidPlacement(cell) else { return false }
        blocks.append(PlacedBlock(col: cell.col, row: cell.row,
                                  designID: designID, isCracked: isCracked, placedAt: date))
        return true
    }

    /// Removes a block. With free placement nothing falls — other blocks stay
    /// exactly where they are.
    @discardableResult
    mutating func removeBlock(at cell: GridCell) -> PlacedBlock? {
        guard let index = blocks.firstIndex(where: { $0.cell == cell }) else { return nil }
        return blocks.remove(at: index)
    }

    /// Where the block at `cell` could be moved: any other empty cell.
    func validMoveDestinations(from cell: GridCell) -> [GridCell] {
        guard isOccupied(cell) else { return [] }
        return validCells.filter { $0 != cell }
    }

    /// Moves a block to any empty in-bounds cell, keeping its identity
    /// (design, cracked state, placement date).
    @discardableResult
    mutating func moveBlock(from: GridCell, to: GridCell) -> Bool {
        guard from != to, let block = block(at: from), isValidPlacement(to) else { return false }
        removeBlock(at: from)
        var moved = block
        moved.col = to.col
        moved.row = to.row
        blocks.append(moved)
        return true
    }

    var builtCount: Int { blocks.filter { !$0.isCracked }.count }
    var crackedCount: Int { blocks.filter(\.isCracked).count }

    /// Maximal horizontal runs of >= minLength contiguous (no-gap) blocks of
    /// the given design in the same row.
    func horizontalRuns(ofDesign designID: String, minLength: Int = 3) -> [BlockRun] {
        var byRow: [Int: [Int]] = [:]
        for b in blocks where b.designID == designID && !b.isCracked {
            byRow[b.row, default: []].append(b.col)
        }
        var runs: [BlockRun] = []
        for (row, unsorted) in byRow {
            let cols = unsorted.sorted()
            var start = cols[0]
            var prev = cols[0]
            for c in cols.dropFirst() {
                if c == prev + 1 { prev = c; continue }
                if prev - start + 1 >= minLength { runs.append(BlockRun(row: row, startCol: start, endCol: prev)) }
                start = c
                prev = c
            }
            if prev - start + 1 >= minLength { runs.append(BlockRun(row: row, startCol: start, endCol: prev)) }
        }
        return runs
    }

    /// Maximal vertical runs of >= minLength contiguous blocks of the given
    /// design in the same column.
    func verticalRuns(ofDesign designID: String, minLength: Int = 3) -> [ColumnRun] {
        var byCol: [Int: [Int]] = [:]
        for b in blocks where b.designID == designID && !b.isCracked {
            byCol[b.col, default: []].append(b.row)
        }
        var runs: [ColumnRun] = []
        for (col, unsorted) in byCol {
            let rows = unsorted.sorted()
            var start = rows[0]
            var prev = rows[0]
            for r in rows.dropFirst() {
                if r == prev + 1 { prev = r; continue }
                if prev - start + 1 >= minLength { runs.append(ColumnRun(col: col, startRow: start, endRow: prev)) }
                start = r
                prev = r
            }
            if prev - start + 1 >= minLength { runs.append(ColumnRun(col: col, startRow: start, endRow: prev)) }
        }
        return runs
    }

    /// Water runs grow a patrolling fish; garden runs grow flowers; a vertical
    /// stack of neon throws a laser party.
    var waterRuns: [BlockRun] { horizontalRuns(ofDesign: "water") }
    var gardenRuns: [BlockRun] { horizontalRuns(ofDesign: "garden") }
    var greenhouseRuns: [BlockRun] { horizontalRuns(ofDesign: "greenhouse") }
    var neonRuns: [ColumnRun] { verticalRuns(ofDesign: "neon") }
}

struct BlockRun: Equatable {
    let row: Int
    let startCol: Int
    let endCol: Int
}

struct ColumnRun: Equatable {
    let col: Int
    let startRow: Int
    let endRow: Int
}

struct ArchivedWorld: Codable, Equatable {
    var archivedAt: Date
    var blocks: [PlacedBlock]
    var name: String? = nil
}

/// On-disk representation of world.json: the live world plus archives
/// from "Start a fresh canvas".
struct WorldFile: Codable, Equatable {
    var current: World = World()
    var archived: [ArchivedWorld] = []
}

File: docs/js/world.js
// The persistent 2D block world — mirrors Models/World.swift.
// Row 0 is the ground (bottom); rows grow upward.

export const COLUMNS = 28;
export const ROWS = 14;

export function cellKey(col, row) {
  return `${col},${row}`;
}

// A world is { blocks: [{ id, col, row, designID, isCracked, placedAt }] }.
export function emptyWorld() {
  return { blocks: [] };
}

function occupancy(world) {
  const set = new Set();
  for (const b of world.blocks) set.add(cellKey(b.col, b.row));
  return set;
}

export function blockAt(world, col, row) {
  return world.blocks.find((b) => b.col === col && b.row === row) || null;
}

export function isOccupied(world, col, row) {
  return blockAt(world, col, row) !== null;
}

export function inBounds(col, row) {
  return col >= 0 && col < COLUMNS && row >= 0 && row < ROWS;
}

// Gravity rule: a block may sit on the ground row or directly on top of an
// occupied cell. No floating blocks.
export function isValidPlacement(world, col, row) {
  if (!inBounds(col, row) || isOccupied(world, col, row)) return false;
  if (row === 0) return true;
  return isOccupied(world, col, row - 1);
}

// Exactly one valid cell per column: the lowest unoccupied cell whose
// support exists (ground or top of the column's stack).
export function validCells(world) {
  const occ = occupancy(world);
  const result = [];
  for (let col = 0; col < COLUMNS; col++) {
    let row = 0;
    while (occ.has(cellKey(col, row))) row++;
    if (inBounds(col, row)) result.push({ col, row });
  }
  return result;
}

export function columnHeight(world, col) {
  let row = 0;
  while (isOccupied(world, col, row)) row++;
  return row;
}

export function mostRecentBlock(world) {
  let best = null;
  for (const b of world.blocks) {
    if (!best || b.placedAt > best.placedAt) best = b;
  }
  return best;
}

export function nearestValidGroundCell(world, col) {
  const ground = validCells(world).filter((c) => c.row === 0);
  if (ground.length === 0) return null;
  return ground.reduce((a, b) => (Math.abs(a.col - col) <= Math.abs(b.col - col) ? a : b));
}

// Default target when the user starts without picking a spot.
export function defaultTarget(world) {
  const recent = mostRecentBlock(world);
  if (recent) {
    const above = { col: recent.col, row: columnHeight(world, recent.col) };
    if (isValidPlacement(world, above.col, above.row)) return above;
    return nearestValidGroundCell(world, recent.col);
  }
  return nearestValidGroundCell(world, Math.floor(COLUMNS / 2));
}

export function placeBlock(world, { designID, isCracked, col, row, placedAt }) {
  if (!isValidPlacement(world, col, row)) return false;
  world.blocks.push({
    id: crypto.randomUUID(),
    col, row, designID, isCracked, placedAt,
  });
  return true;
}

// Remove a block; everything above it in the column falls one row so the
// gravity invariant is preserved.
export function removeBlock(world, col, row) {
  const index = world.blocks.findIndex((b) => b.col === col && b.row === row);
  if (index < 0) return null;
  const [removed] = world.blocks.splice(index, 1);
  for (const b of world.blocks) {
    if (b.col === col && b.row > row) b.row -= 1;
  }
  return removed;
}

export function validMoveDestinations(world, col, row) {
  if (!isOccupied(world, col, row)) return [];
  const clone = cloneWorld(world);
  removeBlock(clone, col, row);
  return validCells(clone).filter((c) => !(c.col === col && c.row === row));
}

export function moveBlock(world, from, to) {
  if (from.col === to.col && from.row === to.row) return false;
  const block = blockAt(world, from.col, from.row);
  if (!block) return false;
  const clone = cloneWorld(world);
  removeBlock(clone, from.col, from.row);
  if (!isValidPlacement(clone, to.col, to.row)) return false;
  // Apply to the clone, then adopt it.
  const moved = { ...block, col: to.col, row: to.row };
  clone.blocks.push(moved);
  world.blocks = clone.blocks;
  return true;
}

export function cloneWorld(world) {
  return { blocks: world.blocks.map((b) => ({ ...b })) };
}

export function builtCount(world) {
  return world.blocks.filter((b) => !b.isCracked).length;
}

export function crackedCount(world) {
  return world.blocks.filter((b) => b.isCracked).length;
}
